### PR TITLE
Oceanwater-1026-BalovnevModelPlugin

### PR DESCRIPTION
**Pull request by imynnnn**
## Linked Issues:
| EPIC ⚡ | [OCEANWATER-XXX](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-XXX) |
| :----------- | :----------- |
| Jira Ticket 🎟️ | [OCEANWATER-1026](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-1026) |
| Github :octocat:  | # |

## Summary of Changes
* Created BalovnevModelPlugin
* Added changes from previous closed PR ( OW-1027-BalovnevModelPlugin #259)
* Updated the correct Jira Ticket

## Test
* Running following command in the Terminal
* ```roslaunch ow europa_terminator_workspace.launch```
* ```tail -F ~/.gazebo/server-11345/default.log```
* ```./grind_action_client.py``` in the ow_lander/scripts folders
* ```./dig_linear_action_client.py``` in the ow_lander/scripts folders

### CODE REVIEW

**Comment by AstroStucky** [comment]
This PR contains both a directory for `BavlovnevModelPlugin` (the mispelled version) and `BalovnevModelPlugin`. Please remove the misspelled directory with a commit on this branch. I would take caution to make sure the suggested changes you made for the previous closed PR are present in the correctly spelled directory before removing the incorrectly spelled directory.

**Comment by AstroStucky** [comment]
There are header files for the new Arm Move Joint(s) actions that **ow_autonomy** is looking for but cannot find during compilation. I suspect it's because this branch needs to be merged with the newest version of noetic-devel. I think you will only see this compilation error if your **ow_autonomy** is on the newest version of **noetic-devel**. 

@imynnnn Please perform the following while in the ow_simulator directory

1. Update noetic-devel to the newest version 
```
git checkout noetic-devel
git pull
```
2. Merge with noetic-devel with this branch
```
git checkout OCEANWATER-1026-BalovnevModelPlugin
git merge noetic-devel
```
3. Hopefully not merge conflicts occur. Then perform a push.
```
git push
```

Thanks!

File: ow_gazebo_plugins/src/BalovnevModelPlugin/BalovnevModelPlugin.h [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+{
+public:
+  
+  void Load(gazebo::physics::ModelPtr model, sdf::ElementPtr sdf);
+  
+private:
+
+  ros::Subscriber m_sub_mod_diff_visual;
+
+  void OnUpdate();
```
Comment (kmdalal) [inline ow_gazebo_plugins/src/BalovnevModelPlugin/BalovnevModelPlugin.h:29]: Our naming convention for member functions would have it be `onUpdate`

File: ow_gazebo_plugins/src/BalovnevModelPlugin/BalovnevModelPlugin.h [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+
+namespace gazebo
+{
+// This plugin is added to a robot description and applies forces calculated 
+// from Balovnev force model
+class BalovnevModelPlugin : public gazebo::ModelPlugin
+{
+public:
+
```
Comment (kmdalal) [inline ow_gazebo_plugins/src/BalovnevModelPlugin/BalovnevModelPlugin.h:22]: It's good to be explicit about the compiler generated functions.  In this case the default ones are being generated, which looks okay.  But you can be clear that this is what you want with the following (my syntax should be very close if not correct).
```suggestion
  BalovnevModelPlugin() = default;
  BalovnevModelPlugin (const BalovnevModelPlugin&) = default;
  BalovnevModelPlugin& operator= (const BalovnevModelPlugin&) = default;
  ~BalovnevModelPlugin() = default;
```

**Review by kmdalal — APPROVED** [review]
I just did a light code review and left some minor suggestions.  I'll leave a more careful review to the others.  Nice work!

**Review by AstroStucky — APPROVED** [review]
Only had a few more code comments this time around. Thank you for incorporating my previous suggested changes.

I would like to give this a thorough test where I compare forces with and without the plugin activated, but it's the end of the week and the end of Yinan's internship, so I'm going to go ahead and approve it is since what testing I did shows that nothing is broken. However, I did have code suggestions, so please take time to look through those and apply where needed.

The framework we have here should in theory work to inject Balovnev forces into the scoop. If there are minor tweaks required, those can be done after Yinan's internship. 

Great work, Yinan! It's been a pleasure working with you and I wish you the best of luck at JSC. :confetti_ball: :smile:

**Comment by imynnnn** [comment]
It has been my pleasure to work on the OCEANWATERS. Thanks for all the support.

**Pull request merged**